### DISCUSSION TITLE
Replacement for from_2d_array_to_nested()

### DISCUSSION THREAD

**Discussion by chaitradhanorkar** [alan-turing-institute/sktime#3385] [Q&A]
Hi, I recently updated my sktime and my old code which uses from_2d_array_to_nested() is not working anymore. I need to format data to feed it to the minirocket model. Can someone please tell me what function I can use to process my data?

**Comment by fkiraly** [comment]
if it is a classifier, it should accept the 2D data without a need for conversion (and do the conversion automatically, internally).

For a manual conversion, you can use `convert` or `convert_to`:
https://www.sktime.org/en/stable/api_reference/utils.html

This tutorial explains the most common formats and how to use the converters:
https://github.com/alan-turing-institute/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb

**Reply by chaitradhanorkar** [reply]
Thank you for this!! I will go through this thoroughly and see what suits my use case.

**Reply by chaitradhanorkar** [reply]
Hi I am still struggling to get numpy.ndarray to a nested dataframe using the current package. 

It give the following error - 
NotImplementedError: no conversion defined from type np.ndarray to nested_univ

**Reply by fkiraly** [reply]
If you have multiple time series in 2D array format, use the string `numpyflat` instead of `np.ndarray`, that should work.
If your array is 3D, use `numpy3D`.

Apologies, the naming is a bit confusing, but stems from historical development and the fact that you can encode one or multiple time series as `numpy` arrays in different ways.

**Reply by chaitradhanorkar** [reply]
Thank you for responding.
Hi, my data is like this - 
<img width="593" alt="image" src="https://user-images.githubusercontent.com/30472037/188697115-17ba9a22-28ce-46da-a596-2d3e62833c3e.png">

and after using the old from_2d_array_to_nested() function I used to get data in this format - 

Xnested = from_2d_array_to_nested(all_X,columns=["rms"])
The shape of this is (n,1)

<img width="286" alt="image" src="https://user-images.githubusercontent.com/30472037/188697263-d2904a58-dc98-4016-932c-bdc37f783523.png">


I tried to use numpyflat too, but it gives me the following issue - 

<img width="530" alt="image" src="https://user-images.githubusercontent.com/30472037/188697434-e3896d99-cdc7-4437-a9ad-a7af21fb0797.png">
NotImplementedError: no conversion defined from type np.ndarray to numpyflat


I can't seem to understand where I am going wrong...

**Reply by chaitradhanorkar** [reply]
So my data is in 'numpyflat'. I think I got what you meant.
<img width="287" alt="image" src="https://user-images.githubusercontent.com/30472037/188698603-4289c697-568e-40f0-bfa9-6bae4fa45024.png">

How do I get it in the nested data frame format with the shape being (n, 1) as I have attached in the previous comment?

**Reply by fkiraly** [reply]
use `convert`, e.g.,
`X_new = convert(all_X, from_type="numpyflat", to_type="nested_univ")`

As said though, sklearn classifiers should be able to work with data in `numpyflat` format directly.

**Reply by chaitradhanorkar** [reply]
Thank you! I was later trying to convert from "np.ndarray" to "nested_univ" which was causing issues.  "numpyflat" to "nested_univ" works. sktime MiniRocket Univariate was not accepting the dimensions of my np.ndarray. It needed (n,1), hence the need for "nested_univ".

**Reply by fkiraly** [reply]
yes, transformers don't accept `numpyflat`, only classifiers.
Transformers accept `numpy3D` though, in your case that would be (n,1,m).
